---
title: "Stacked Violin Plot Atlantic Salmon"
author: "Sarah Salisbury"
output:
  html_document:
    toc: true
    number_sections: true
---

# Getting Your Data

## Setting up R Environment

### Setup
```{r setup}
# Prevent scientific notation output (e.g., we don't want very small p-values to be output as say 10^-9)
options(scipen = 999)
#set seed for reproducibility
set.seed(42) #note that UMAPs use seed of 42, not sure if setting seed here overrides this.
```

### Load Packages

Now load up all of the packages you'll need for this analysis

```{r load packages}
# Load up libraries
library(Seurat) # to run single cell analyses
library(data.table)
library(dplyr)
library(ggplot2)
library(cowplot)
library(patchwork)
library(stringr)
```

### Load the Seurat object
```{r load Seurat object}
AS.integrated <- readRDS("../../ATLANTIC_SALMON/4_Filter1_Incorporate_Immune_Cell_Substructuring/AS.integrated_filter1_subset.rds")
```

```{r set default assay}
# For performing differential expression after integration, we switch back to the SCT data: https://satijalab.org/seurat/articles/sctransform_v2_vignette.html#perform-integration-using-pearson-residuals-1
DefaultAssay(AS.integrated) <- "SCT"
```

```{r set clusters}
# Set identity classes to an existing column in meta data
Idents(object = AS.integrated) <- "SUBCLUSTER_MERGED" # the column where we broke down immune and cluster 12 cells
```

```{r stacked bar plots}
#https://satijalab.org/seurat/articles/pbmc3k_tutorial.html
new.cluster.ids <- c("1" = "Fibroblast_1",
                     "2" = "Intermediate_Keratinocytes",
                     "3" = "Basal_Keratinocytes",
                     "5" = "Secretory",
                     "6" = "Mucous", 
                     "7" = "Osteoblasts",
                     "8" = "Endothelial",
                     "9" = "Glial_1",
                     "10" = "Keratinocyte_Suprabasal",
                     "11" = "Red_Blood_Cell",
                     "12" = "Neuronal_1",
                     "13" = "Ltk_cells",
                     "14" = "Fibroblast_2",
                     "16" = "Undifferentiated",
                     "17" = "Glial_2",
                     "18" = "Muscle_cells",
                     "20" = "Neuronal_2",
                     "Dendritic_cells_5" = "Dendritic_cells_5",
                     "Myeloid" = "Myeloid",
                     "T_cells" = "T_cells",
                     "B_cells_10" = "B_cells_10",
                     "Neutrophil_cells_7" = "Neutrophil_cells_7",
                     "Monocyte_11" = "Monocyte_11"
                     )


#https://github.com/satijalab/seurat/issues/3500
AS.integrated <- RenameIdents(AS.integrated, new.cluster.ids)
AS.integrated$CellTypeNames <- Idents(AS.integrated) # save in metadata
```

```{r}
cluster.id.plot.labels <- c("Fibroblast_1" = "Fibroblasts (1)",
                            "Intermediate_Keratinocytes" = "Intermediate Keratinocytes",
                            "Basal_Keratinocytes" = "Basal Keratinocytes",
                            "Endothelial" = "Endothelial", 
                            "Mucous" = "Mucous",
                            "Osteoblasts" = "Osteoblasts",
                            "Secretory" = "Secretory",
                            "Glial_1" = "Glial (1)",
                            "Keratinocyte_Suprabasal" = "Suprabasal Keratinocytes",
                            "Red_Blood_Cell" = "Red Blood Cells",
                            "Neuronal_1" = "Neuronal (1)",
                            "Ltk_cells" = "Neural Crest Cells",
                            "Fibroblast_2" = "Fibroblasts (2)",
                            "Undifferentiated" = "Undifferentiated",
                            "Glial_2" = "Glial (2)",
                            "Muscle_cells" = "Muscle",
                            "Neuronal_2" = "Neuronal (2)",
                            "Dendritic_cells_5" = "Dendritic Cells",
                            "Myeloid" = "Macrophages",
                            "T_cells" = "T Cells",
                            "B_cells_10" = "B Cells",
                            "Neutrophil_cells_7" = "Neutrophils",
                            "Monocyte_11" = "Monocyte")
```

```{r}
order_plots <- c("T_cells",
                 "B_cells_10",
                 "Dendritic_cells_5",
                 "Neutrophil_cells_7",
                 "Myeloid",
                 "Monocyte_11",
                 "Keratinocyte_Suprabasal",
                 "Intermediate_Keratinocytes",
                 "Basal_Keratinocytes",
                 "Fibroblast_1",
                 "Fibroblast_2",
                 "Endothelial",
                 "Secretory",
                 "Mucous",
                 "Osteoblasts",
                 "Red_Blood_Cell",
                 "Muscle_cells",
                 "Ltk_cells",
                 "Neuronal_1",
                 "Neuronal_2",
                 "Glial_1",
                 "Glial_2",
                 "Undifferentiated")

# Don't want to remake factor according to: https://github.com/satijalab/seurat/issues/3500
# Make Identity factor orders
#AS.integrated$CellTypeNames <- factor(AS.integrated$CellTypeNames, levels = order_plots)
#Idents(AS.integrated) <- AS.integrated$CellTypeNames

#note that we already switched our identities to those in column CellTypeNames, so by setting the levels to "order_plots" which is just a rearrangement of those levels already present in CellTypeNames this is totally fine
#orders your cell type names by the order you specified in order_plots
levels(AS.integrated) <- order_plots # note this is ok to do for a Seurat object but not ok for levels in a factor in a column! As the latter will rename all your levels !!!
```

```{r}
markers <- c("SWAP70.1" = "B_cells_10",
             "tbc1d10c" = "T_cells",
             "csf3r.1" = "Neutrophil_cells_7",
             "mmp9" = "Neutrophil_cells_7",
             "mmp13" = "Neutrophil_cells_7",
             "flt3" = "Dendritic_cells_5",
             "sema4ab.1" = "Myeloid",
             "mitfa" = "Monocyte_11",
             "ckb" = "Monocyte_11",
             "fbn2b" = "1",
             "ass1" = "2",
             "ITGA6" = "3",
             "lamb4" = "3",
             "pcdh11" = "5",
             "fer1l4" = "6",
             "panx3" = "7",
             "itga10" = "7",
             "egfl7" = "8",
             "abcg1" = "9",
             "evpla" = "10", 
             "HBB1" = "11",
             "nwd1" = "12",
             "ltk" = "13",
             "mlphb" = "13",
             "lamc1" = "14",
             "gene:ENSSSAG00000006052" = "16", #pax7
             "col4a1" = "16",
             "WDR49" = "17",
             "casq1b.1" = "18",
             "SAMD12.1" = "20",
             "gene:ENSSSAG00000038884" = "20" #il10rbl
)
             
markernames <- names(markers)             
markergroups <- unname(markers) #https://www.tutorialspoint.com/how-to-remove-names-from-a-named-vector-in-r             

features <- as.data.frame(cbind(markernames,markergroups), stringsAsFactors = FALSE)     #bind together the marker names and groups into a single dataframe         
features <- features %>% mutate(nicemarkergroups = unname(new.cluster.ids[markergroups])) # now take the markergroup value in each column and look up it's value in the new.cluster.ids named cluster to give you the nice name of the cluster, then unname it (to remove the name e.g., 1, 2, 3, 9) and save in a new column "nicemarkergroups"

colours_figs <- c("T_cells"  = "darkorchid4",
               "Myeloid" = "magenta2",
               "Dendritic_cells_5" = "orchid1",
               "Neutrophil_cells_7" = "hotpink",
               "Monocyte_11" = "darkmagenta",
               "B_cells_10" = "mediumorchid",
               "Intermediate_Keratinocytes" = "darkgoldenrod",
               "Basal_Keratinocytes" = "darkgoldenrod4",
               "Keratinocyte_Suprabasal" = "darkgoldenrod2",
               "Fibroblast_1" = "chocolate3",
               "Secretory" = "deepskyblue4",
               "Endothelial" = "deepskyblue",
               "Mucous" = "darkolivegreen",
               "Osteoblasts" = "burlywood1",
               "Red_Blood_Cell" = "firebrick1",
               "Muscle_cells" = "darkred",
               "Ltk_cells" = "lightblue3",
               "Neuronal_1" = "slateblue3",
               "Neuronal_2" = "navy",
               "Glial_1" = "thistle",
               "Glial_2" = "thistle4",
               "Fibroblast_2" = "khaki3",
               "Undifferentiated" = "salmon"
)

features <- features %>% mutate(colour = unname(colours_figs[nicemarkergroups])) # now take the markergroup value in each column and look up it's value in the colours_figs named cluster to give you the colour for that cluster, then unname it (to remove the name e.g. "Glial") and save in a new column "colour"

features <- features[order(factor(features$nicemarkergroups, levels = rev(order_plots))),] # https://stackoverflow.com/questions/34749363/sort-data-frame-in-specific-order-not-ascending-descending

b_AS <- VlnPlot(AS.integrated, assay = "SCT", slot = "data", features$markernames, stack = TRUE, sort = FALSE, flip = TRUE, cols = features$colour) +
        theme(legend.position = "none",
              panel.spacing = unit(0, "lines"),
              panel.background = element_rect(fill = NA, color = "black"),
              strip.background = element_blank(),
              strip.text = element_text(face = "bold"),
              strip.text.y.left = element_text(angle = 0),
              axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
              axis.title.x = element_blank()) +
  scale_x_discrete(labels = cluster.id.plot.labels) # make nice labels
```
```{r}
df_colours_AS <-data.frame(x = order_plots,
                colour = c("darkorchid4", # T_cells
                           "mediumorchid", # B_cells
                           "orchid1", # Dendritic_cells
                           "hotpink", # Neutrophil_cells
                           "magenta2", # Myeloid
                           "darkmagenta", #Monocyte
                           "darkgoldenrod2", #Keratinocyte_Suprabasal
                           "darkgoldenrod", # Intermediate_Keratinocytes
                           "darkgoldenrod4", #Basal_Keratinocytes
                           "chocolate3", #Fibroblast_1
                           "khaki3", #Fibroblast_2
                           "deepskyblue", #Endothelial
                           "deepskyblue4", #Secretory
                           "darkolivegreen", #Mucous
                           "burlywood1", #Osteoblasts
                           "firebrick1", #Red_Blood_Cell
                           "darkred", #Muscle_cells
                           "lightblue3", #Ltk_cells
                           "slateblue3", #Neuronal_1
                           "navy", #Neuronal_2
                           "thistle", #Glial_1
                           "thistle4", #Glial_2
                           "salmon" #Undifferentiated
                ))

df_colours_AS$x <- factor(df_colours_AS$x, levels = df_colours_AS$x)

plotAS <- ggplot(df_colours_AS, aes(x=x, y=1, fill = x)) +
  geom_tile() +
  theme_void() +
  scale_fill_manual(values=df_colours_AS$colour) +
  theme(legend.position = 'none') +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, margin = margin(t = 10), size = 20)) +
  scale_x_discrete(labels = cluster.id.plot.labels) + # make nice labels
  theme(plot.margin = unit(c(0, 0, 0, 0), "cm"))


b_AS_noaxis <-
  b_AS + 
  theme(axis.text.x=element_blank())

# Now let's combine the proportion plots for each cluster and for all samples:
b_AS_nice <- plot_grid(b_AS_noaxis,
                       plotAS, # specify proportion plot for all samples
                                                          ncol = 1, # plot one on top of the other with 1 column
                                                          rel_heights = c(3, 1.2), # specify relative heights
                                                          align = "v", # align vertically
                                                          axis = "lr") # align at left and right axes

b_AS_nice

# Use plot_grid to join plots
pdf("markers_AS_nice.pdf", width = 10, height = 13, useDingbats = FALSE)
plot_grid(b_AS_nice)
dev.off()
```

